runtime/src/accounts_db.rs

```diff
@@ -3249,6 +3249,172 @@ impl AccountsDb {
+        let mut stored_accounts = stored_accounts.into_iter().collect::<Vec<_>>();
+        stored_accounts.sort_unstable_by(|a, b| a.0.cmp(&b.0));
+
+        let keep_accounts_collect = Mutex::new(Vec::with_capacity(stored_accounts.len()));
+        let remove_pubkeys_collect = Mutex::new(Vec::with_capacity(stored_accounts.len()));
+        let total_bytes_collect = AtomicUsize::new(0);
+        const CHUNK_SIZE: usize = 50;
+        stored_accounts.par_chunks(CHUNK_SIZE).for_each(|chunk| {
+            let mut chunk_bytes = 0;
+            let mut keep_accounts = Vec::with_capacity(CHUNK_SIZE);
+            let mut remove_pubkeys = Vec::with_capacity(CHUNK_SIZE);
+            chunk.iter().for_each(|(pubkey, account)| {
+                if minimized_account_set.contains(pubkey) {
+                    chunk_bytes += account.account_size;
+                    keep_accounts.push((pubkey, account));
+                } else if self.accounts_index.get_account_read_entry(pubkey).is_some() {
+                    remove_pubkeys.push(pubkey);
+                }
+            });
+
+            keep_accounts_collect
+                .lock()
+                .unwrap()
+                .append(&mut keep_accounts);
+            remove_pubkeys_collect
+                .lock()
+                .unwrap()
+                .append(&mut remove_pubkeys);
+            total_bytes_collect.fetch_add(chunk_bytes, Ordering::Relaxed);
+        });
+
+        let keep_accounts = keep_accounts_collect.into_inner().unwrap();
+        let remove_pubkeys = remove_pubkeys_collect.into_inner().unwrap();
+        let total_bytes = total_bytes_collect.load(Ordering::Relaxed);
+
+        let remove_pubkeys: Vec<_> = remove_pubkeys
+            .into_iter()
+            .map(|pubkey| (*pubkey, slot))
+            .collect();
+        self.purge_keys_exact(remove_pubkeys.iter());
+
+        let aligned_total: u64 = Self::page_align(total_bytes as u64);
+        let dead_storages_original_length = dead_storages.len();
+        if aligned_total > 0 {
+            let mut accounts = Vec::with_capacity(keep_accounts.len());
+            let mut hashes = Vec::with_capacity(keep_accounts.len());
+            let mut write_versions = Vec::with_capacity(keep_accounts.len());
+
+            for (pubkey, alive_account) in keep_accounts {
+                accounts.push((pubkey, &alive_account.account));
+                hashes.push(alive_account.account.hash);
+                write_versions.push(alive_account.account.meta.write_version);
+            }
+
+            let (new_storage, _time) = self.get_store_for_shrink(slot, aligned_total);
+
+            self.store_accounts_frozen(
+                (slot, &accounts[..]),
+                Some(&hashes),
+                Some(&new_storage),
+                Some(Box::new(write_versions.into_iter())),
+            );
+
+            new_storage.flush().unwrap();
+        }
+
+        self.shrink_candidate_slots.lock().unwrap().remove(&slot);
```

Comment:
how is `shrink_candidate_slots` used where we have to do this here? Do we end up calling shrink somehow?